docs/source/en/tutorials/passport.md

```diff
@@ -1 +1,282 @@
+
+```js
+// app.js
+Module.exports = app => {
+  App.passport.verify(async (ctx, user) => {
+    // Check user
+    Assert(user.provider, 'user.provider should exists');
+    Assert(user.id, 'user.id should exists');
+
+    // Find user information from the database
+    //
+    // Authorization Table
+    // column   | desc
+    // ---      | --
+    // provider | provider name, like github, twitter, facebook, weibo and so on
+    // uid      | provider unique id
+    // user_id  | current application user id
+    Const auth = await ctx.model.Authorization.findOne({
+      Uid: user.id,
+      Provider: user.provider,
+    });
+    Const existsUser = await ctx.model.User.findOne({ id: auth.user_id });
+    If (existsUser) {
+      Return existsUser;
+    }
+    // Call service to register a new user
+    Const newUser = await ctx.service.user.register(user);
+    Return newUser;
+  });
+
+  // Serialize and store the user information into session. Generally, only a few fields need to be streamlined/saved.
+  app.passport.serializeUser(async (ctx, user) => {
+  // process user
+  // ...
+  // return user;
+  });
+
+  // Deserialize the user information from the session, check the database to get the complete information
+  app.passport.deserializeUser(async (ctx, user) => {
+  // process user
+  // ...
+  // return user;
+  });
+};
+```
+
+At this point, we have completed all the configurations. For a complete example, see: [eggjs/examples/passport](https://github.com/eggjs/examples/tree/master/passport)
+
+### API
+
+[egg-passport](https://github.com/eggjs/egg-passport) provides the following extensions:
+
+* `ctx.user` - Get currently logged in user information
+* `ctx.isAuthenticated()` - Check if the request is authorized
+* `ctx.login(user, [options])` - Start a login session for the user
+* `ctx.logout()` - Exit and clear user information from session
+* `ctx.session.returnTo=` - Set redirect address after success before jumping to authentication
+
+The API also provides:
+
+* `app.passport.verify(async (ctx, user) => {})` - Check user
+* `app.passport.serializeUser(async (ctx, user) => {})` - Serialize user information into session
+* `app.passport.deserializeUser(async (ctx, user) => {})` - Deserialize user information from the session
+* `app.passport.authenticate(strategy, options)` - Generate the specified authentication middleware
+    - `options.successRedirect` - specifies the redirect address after successful authentication
+    - `options.loginURL` - jump login address, defaults to `/passport/${strategy}`
+    - `options.callbackURL` - callback address after authorization, defaults to `/passport/${strategy}/callback`
+* `app.passport.mount(strategy, options)` - Syntactic sugar for developers to configure routing
+
+## Using Passport Ecosystem
+
+[Passport](http://www.passportjs.org/) has many middleware and it is impossible to carry out secondary packaging.
```

Comment:
impossible to carry out secondary packaging -> impossible to have the second encapsulation